File: ts/jsonDiff.ts
const getType = (value) => Object.prototype.toString.call(value).slice(8, -1).toLowerCase();
export function diff (oldObj: any, newObj: any): any {
	if (newObj === oldObj) {
		return undefined;
	}
	let oldObjType = getType(oldObj);
	let newObjType = getType(newObj);
	if (newObjType == oldObjType) {
		if (newObjType == "array") {
			let result = {};
			for (let index in newObj) {
				let newValue = diff(oldObj[index], newObj[index]);
				if (newValue !== undefined) {
					result[index] = newValue;
				}
			}
			if (newObj.length == oldObj.length && Object.keys(result).length == 0) {
				return undefined;
			}
			result["$length"] = newObj.length;
			return result;
		}
		else if (newObjType == "object") {
			let result = {};
			for (let index in newObj) {
				let newValue = diff(oldObj[index], newObj[index]);
				if (newValue !== undefined) {
					result[index] = newValue;
				}
			}
			let unset: string[] = [];
			for (let index in oldObj) {
				if (newObj[index] === undefined) {
					unset.push(index);
				}
			}
			if (unset.length > 0) {
				result["$unset"] = unset;
			}
			if (Object.keys(result).length == 0) {
				return undefined;
			}
			return result;
		}
	}
	return newObj;
}
export function apply (oldObj: any, diffObj: any): any {
	let oldObjType = getType(oldObj);
	let diffObjType = getType(diffObj);
	if (diffObjType == "object" && diffObj["$length"] !== undefined) {
		diffObjType = "array";
		let length = diffObj["$length"];
		if (diffObjType == oldObjType) {
			while (oldObj.length < length) {
				oldObj.push(undefined);
			}
			if (oldObj.length > length) {
				oldObj.splice(length, oldObj.length - length);
			}
			for (let index in diffObj) {
				if (index.indexOf("$") != 0) {
					oldObj[index] = apply(oldObj[index], diffObj[index]);
				}
			}
			return oldObj;
		}
		else {
			let result: any[] = [];
			while (result.length < length) {
				result.push(undefined);
			}
			for (let index in diffObj) {
				if (index.indexOf("$") != 0) {
					result[index] = diffObj[index];
				}
			}
			return result;
		}
		
	}
	else if (diffObjType == "object") {
		if (diffObjType == oldObjType) {
			let unset = diffObj["$unset"];
			if (unset !== undefined) {
				for (let i = 0; i < unset.length; i++) {
					delete oldObj[unset[i]];
				}
			}
			for (let index in diffObj) {
				if (index.indexOf("$") != 0) {
					oldObj[index] = apply(oldObj[index], diffObj[index]);
				}
			}
			return oldObj;
		}
		else {
			for (let index in diffObj) {
				if (index.indexOf("$") == 0) {
					delete diffObj[index];
				}
			}
			return diffObj;
		}
	}
	return diffObj;
}

File: cs/JsonDiff.cs
using UnityEngine;
using System;
using System.Collections.Generic;

public static class JsonDiff {
	
	static string TypeNameLower<T>(T obj) {
		return typeof(T).Name.ToLower();
	}
	static string TypeNameLower(object obj) {
	if (obj != null) { return obj.GetType().Name.ToLower(); }
		else { return null; }
	}
	public static object Apply (object oldObj, object diffObj) {
		string oldObjType = TypeNameLower (oldObj);
		string diffObjType = TypeNameLower (diffObj);
		Dictionary<string, object> diffObjDict = diffObj as Dictionary<string, object>;
		if (diffObjType == TypeNameLower (new Dictionary<string, object> ()) && diffObjDict.ContainsKey ("$length")) {
			diffObjType = TypeNameLower (new List<object> ());
			int length = int.Parse (diffObjDict["$length"].ToString ());
			if (diffObjType == oldObjType) {
				List<object> oldObjList = oldObj as List<object>;
				while (oldObjList.Count < length) {
					oldObjList.Add (null);
				}
				if (oldObjList.Count > length) {
					oldObjList.RemoveRange (length, oldObjList.Count - length);
				}
				foreach (var kvp in diffObjDict) {
					if (kvp.Key.IndexOf ("$") != 0) {
						int index = int.Parse (kvp.Key);
						oldObjList[index] = Apply (oldObjList[index], kvp.Value);
					}
				}
				return oldObjList;
			}
			else {
				List<object> result = new List<object> ();
				while (result.Count < length) {
					result.Add (null);
				}
				foreach (var kvp in diffObjDict) {
					if (kvp.Key.IndexOf ("$") != 0) {
						int index = int.Parse (kvp.Key);
						result[index] = kvp.Value;
					}
				}
				return result;
			}
		}
		else if (diffObjType == TypeNameLower (new Dictionary<string, object> ())) {
			if (diffObjType == oldObjType) {
				Dictionary<string, object> oldObjDict = oldObj as Dictionary<string, object>;
				object deletes = null;
				if (diffObjDict.TryGetValue ("$deletes", out deletes)) {
					List<object> deleteList = deletes as List<object>;
					foreach (object deleteObj in deleteList) {
						oldObjDict.Remove (deleteObj.ToString ());
					}
				}
				foreach (var kvp in diffObjDict) {
					if (kvp.Key.IndexOf ("$") != 0) {
						oldObjDict[kvp.Key] = Apply (oldObjDict.ContainsKey (kvp.Key) ? oldObjDict[kvp.Key] : null, kvp.Value);
					}
				}
				return oldObjDict;
			}
			else {
				List<string> deleteKeys = new List<string> ();
				foreach (var kvp in diffObjDict) {
					if (kvp.Key.IndexOf ("$") == 0) {
						deleteKeys.Add (kvp.Key);
					}
				}
				foreach (string deleteKey in deleteKeys) {
					diffObjDict.Remove (deleteKey);
				}
				return diffObjDict;
			}
		}
		return diffObj;
	}
}
